### PR TITLE
Adding environment variables and arguments for process events.

### PR DESCRIPTION
**Pull request by mofarrell**
This pull request make several major changes to kernel communications.  Events now specify a size allowing them to add on a variable size member.  This is now used by the process events in order to support passing of argv and envv to userspace with each event.

### CODE REVIEW

File: kernel/src/circular_queue_kern.c [hunk 1]
```diff
@@ -225,13 +227,9 @@ void *osquery_cqueue_reserve(osquery_cqueue_t *queue, osquery_event_t event) {
...
   osquery_data_header_t *header = NULL;
 
-  ssize_t event_size = osquery_sizeof_event(event);
-  size_t size = event_size;
-  if (event_size >= 0) {
-    // Event exist.  If the event doesnt exist we will try to allocate
-    // a too large amount of space and the allocation will fail.
-    size += sizeof(osquery_data_header_t);
-  }
+  // Event exist.  If the event doesnt exist we will try to allocate
```
Comment (theopolis) [inline kernel/src/circular_queue_kern.c]: sp, event exists. & doesn't

File: kernel/src/circular_queue_kern.c [hunk 2]
```diff
@@ -225,13 +227,9 @@ void *osquery_cqueue_reserve(osquery_cqueue_t *queue, osquery_event_t event) {
...
+  // a too large amount of space and the allocation will fail.
+  size += sizeof(osquery_data_header_t);
```
Comment (theopolis) [inline kernel/src/circular_queue_kern.c]: Possible double count of size of struct.

File: kernel/src/publishers/darwin/process_events.c [hunk 3]
```diff
@@ -12,90 +12,164 @@
 #include <sys/systm.h>
 #include <sys/kauth.h>
 #include <sys/vnode.h>
+#include <sys/imgact.h>
+#include <sys/proc.h>
+#define CONFIG_MACF 1
```
Comment (theopolis) [inline kernel/src/publishers/darwin/process_events.c]: Add a quick doc if this is needed and that it comes from BSD/MAC

File: kernel/src/publishers/darwin/process_events.c [hunk 4]
```diff
@@ -12,90 +12,164 @@
...
+                                             size_t macpolicyattrlen,
+                                             int *disjointp) {
+  int path_len = MAXPATHLEN;
+
+  if (!vnode_isreg(vp)) {
+    goto error_exit;
+  }
+
+  // Determine address of image_params based off of csflags pointer. (HACKY)
+  struct image_params *img;
```
Comment (theopolis) [inline kernel/src/publishers/darwin/process_events.c]: Just initialize.

File: kernel/tools/deployment/Info.plist [hunk 5]
```diff
@@ -31,7 +31,9 @@
 		<key>com.apple.kpi.iokit</key>
 		<string>14.0</string>
 		<key>com.apple.kpi.mach</key>
-		<string>14.0</string>
+                <string>14.0</string>
```
Comment (theopolis) [inline kernel/tools/deployment/Info.plist]: one more space

File: osquery/events/kernel/circular_queue_user.cpp [hunk 6]
```diff
@@ -76,12 +76,11 @@ osquery_event_t CQueue::dequeue(CQueue::event **event) {
   }
   header = (osquery_data_header_t *)read_;
   if (header->event != END_OF_BUFFER_EVENT) {
-    size_t size = osquery_sizeof_event(header->event)
-      + sizeof(osquery_data_header_t);
+    size_t size = header->size;
```
Comment (theopolis) [inline osquery/events/kernel/circular_queue_user.cpp]: Could just stick header->size into the next statement and removal the local size_t.

File: osquery/tables/events/process_events.cpp [hunk 7]
```diff
@@ -38,6 +38,33 @@ Status ProcessEventSubscriber::Callback(
     const TypedKernelEventContextRef<osquery_process_event_t> &ec,
     const void *user_data) {
   Row r;
+
+  r["argc"] = BIGINT(ec->event.argc);
+  r["actual_argc"] = BIGINT(ec->event.actual_argc);
+  r["arg_length"] = BIGINT(ec->event.arg_length);
+  r["envc"] = BIGINT(ec->event.envc);
+  r["actual_envc"] = BIGINT(ec->event.actual_envc);
+  r["env_length"] = BIGINT(ec->event.env_length);
```
Comment (theopolis) [inline osquery/tables/events/process_events.cpp]: add a v

File: osquery/tables/events/process_events.cpp [hunk 8]
```diff
@@ -38,6 +38,33 @@ Status ProcessEventSubscriber::Callback(
     const TypedKernelEventContextRef<osquery_process_event_t> &ec,
     const void *user_data) {
   Row r;
+
+  r["argc"] = BIGINT(ec->event.argc);
+  r["actual_argc"] = BIGINT(ec->event.actual_argc);
+  r["arg_length"] = BIGINT(ec->event.arg_length);
```
Comment (theopolis) [inline osquery/tables/events/process_events.cpp]: add a v

**Comment by mofarrell** [comment]
ok to test

File: osquery/tables/events/process_events.cpp (hunk 8 above)
Comment (marpaia) [inline osquery/tables/events/process_events.cpp]: is there any particular reason why these are bigint instead of int?

**Comment by mofarrell** [comment]
ok to test

File: kernel/src/circular_queue_kern.c [hunk 9]
```diff
@@ -215,23 +215,20 @@ int osquery_cqueue_dropped_data(osquery_cqueue_t *queue) {
...
 
-  ssize_t event_size = osquery_sizeof_event(event);
-  size_t size = event_size;
-  if (event_size >= 0) {
-    // Event exist.  If the event doesnt exist we will try to allocate
-    // a too large amount of space and the allocation will fail.
-    size += sizeof(osquery_data_header_t);
-  }
+  size_t contents_size = size;
+  size += sizeof(osquery_data_header_t);
```
Comment (theopolis) [inline kernel/src/circular_queue_kern.c]: A call to `osquery_cqueue_reserve` will add a `sizeof(...process_t) + ...`. Meaning the caller is still responsible for including the size of it's custom/specific struct. The `sizeof(osquery_data_header_t)` is then added.

Comment (mofarrell) [inline kernel/src/circular_queue_kern.c]: Were you going somewhere with this or is this just a note?

**Pull request merged**